### ISSUE TITLE
Importmap issue with Deno React vite

### ISSUE THREAD

**Issue by dong-lufei** [denoland/deno#17536]
https://deno.land/manual@v1.30.0/node/how_to_with_npm/react

I followed the documentation, deno showed a syntax error
`Relative import path "react" not prefixed with / or ./ or ../deno(import-prefix-missing)`

Should I configure ImportMap in deno.json and fill in vite.config.mjs?

**Comment by bartlomieju** [comment]
Yes, you need to either add `importMap` to `deno.json` or specify it directly in `deno.json` (https://deno.com/blog/v1.30#denojson-becomes-an-import-map).

**Comment by sobanieca** [comment]
I may be wrong, but I think there may be some issue related to `deno install` which may be related to this one as well. I've just migrated my OSS to use `deno.json` `imports` feature (https://github.com/sobanieca/jsonr/commit/f5a397d45624c7b337a55643802a218a3dac76af#diff-54337bcff4b919f6b8b4e0b0e51d06fd9124562fde88899be726c24b7e354b7c) and when I run it locally all is good: 
![image](https://user-images.githubusercontent.com/2606245/215204086-55787722-1732-432e-9336-9aa763c40ec2.png)

I have released it to deno 3rd party repository and when I try to install it I get:

![image](https://user-images.githubusercontent.com/2606245/215204346-673a943e-812a-4ec7-b0b9-4ea74c467e82.png)

Maybe it's something on my end, but maybe `deno install` has some issues with import maps handling?

**Comment by sobanieca** [comment]
Ok, actually I guess the reason why my case is not working is related to the fact that `deno install` most likely doesn't try to find and download `deno.json` file?

**Comment by bartlomieju** [comment]
> Ok, actually I guess the reason why my case is not working is related to the fact that `deno install` most likely doesn't try to find and download `deno.json` file?

Correct, it needs to be explicitly specified as an argument (we can't find config file for remote modules)

**Comment by happyhunter7** [comment]
I'm having now the same issue
let's say I have folder A which is a deno project and has a deno.json file
and then I have project B which also have a deno.json file

Now in project A I'm trying to import a file from project B and that file in project B has (import react from 'react')

Now I have in both projects in deno.json in imports this `"@types/react-dom": "npm:@types/react-dom@18.3.0"`
But even that doesn't help and I get the error that it is a relative import and react needs to be prefixed

How do I fix that @bartlomieju @lucacasonato

**Comment by sobanieca** [comment]
In my case I have given up on using import maps. I've went back to `deps.js`. IMHO it's better solution - it works as expected according to ES modules spec. Even for big projects, I believe it should be fine (it shouldn't be hard to create a lint rule that ensures that `http://` import is done only inside `deps.js` file).

**Comment by happyhunter7** [comment]
Yes, I feel like this is the way to go
I also noticed that if for example module A uses module B then the deps of module B will be not autodownloaded if just using a deno.json file, so I think I also have to use a deps.ts solution

**Comment by bartlomieju** [comment]
Closing — the original problem is resolved.

The `Relative import path "react" not prefixed... (import-prefix-missing)` error is just the bare-specifier-without-a-map case, and the documented fix works today: add the dependency to `"imports"` in `deno.json`. Verified on a current build:

```jsonc
// deno.json
{ "imports": { "react": "npm:react@18.3.1" } }
```
```ts
import React from "react";
console.log(React.version); // 18.3.1
```

The old `deno.land/manual@v1.30` React+Vite page this linked to is gone, and React/Vite-with-npm works fine now.

The only lingering point in the thread (project A importing a file from project B not picking up B's own `deno.json` imports) is by design: a program resolves bare specifiers against a single (root) import map, so a dependency's own `deno.json` imports don't apply to its files when pulled in by another project. To share code with its own dependency resolution, publish it as a JSR/npm package or use a workspace. That's a separate topic from this issue, so closing here.
